atox/src/main/kotlin/ui/user_profile/UserProfileFragment.kt

```diff
@@ -155,23 +170,101 @@ class UserProfileFragment : BaseFragment<FragmentUserProfileBinding>(FragmentUse
-        for (x in 0 until qrData.size) {
-            for (y in 0 until qrData.size) {
-                bmp.setPixel(x, y, if (qrData.getModule(x, y)) Color.BLACK else Color.WHITE)
-            }
-        }
-        val metrics = resources.displayMetrics
-        val size = (min(metrics.widthPixels, metrics.heightPixels) * QR_CODE_TO_SCREEN_RATIO).roundToInt()
-        bmp = bmp.scale(size, size, false)
+        val bmp = createQrCode()
         val qrCode = ImageView(requireContext()).apply {
-            setPadding(dpToPx(QR_CODE_DIALOG_PADDING, resources))
+            setPadding(qrCodePadding.asPx(resources).px)
             setImageBitmap(bmp)
         }
+
         return AlertDialog.Builder(requireContext())
             .setTitle(R.string.tox_id)
             .setView(qrCode)
+            .setPositiveButton(getString(R.string.share)) { _, _ ->
+                vm.viewModelScope.launch {
+                    val qrImageUri = getQrForSharing("tox_id_qr_code")
+                    val shareIntent = Intent().apply {
+                        action = Intent.ACTION_SEND
+                        clipData = ClipData.newRawUri(null, qrImageUri)
+                        type = "image/png"
+                        putExtra(Intent.EXTRA_STREAM, qrImageUri)
+                        flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
+                    }
+                    startActivity(Intent.createChooser(shareIntent, getString(R.string.tox_id_share)))
+                }
+            }
+            .setNegativeButton(getString(R.string.close), null)
             .create()
     }
+
+    /*
+     * Function will save the image with the input bitmap and name to the cache directory as a png format.
+     * Then it will return the Uri of the image.
+     */
+    private fun saveImageForSharing(context: Context, image: Bitmap, name: String): Uri {
+        val imagesFolder = File(context.cacheDir, "shared_images")
+        val uri: Uri
+
+        imagesFolder.mkdirs()
+        val file = File(imagesFolder, "$name.png")
+        FileOutputStream(file).use { stream ->
+            image.compress(Bitmap.CompressFormat.PNG, 90, stream)
+            stream.flush()
+        }
+        uri = FileProvider.getUriForFile(context, "${BuildConfig.APPLICATION_ID}.fileprovider", file)
+
+        return uri
+    }
+
+    // Function will create and return the QR code bitmap of the Tox ID.
+    private fun createQrCode(
+        qrSize: Size = Px(
+            (
+                min(
+                    resources.displayMetrics.widthPixels,
+                    resources.displayMetrics.heightPixels
+                ) * QR_CODE_TO_SCREEN_RATIO
+                ).toInt()
+        ),
+        padding: Size = qrCodePadding
+    ): Bitmap {
+        // Creating the QR bitmap
+        val qrData = QrCode.encodeText("tox:%s".format(vm.toxId.string()), QrCode.Ecc.LOW)
+        val qrCodeColor: Int = Color.BLACK
+        val backgroundColor: Int = Color.WHITE
```

Comment:
It's probably easier to follow the code if you inline the colors where they're used instead of using variables for them.